Lib/subprocess.py

```diff
@@ -1480,7 +1480,26 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
             if shell:
                 startupinfo.dwFlags |= _winapi.STARTF_USESHOWWINDOW
                 startupinfo.wShowWindow = _winapi.SW_HIDE
-                comspec = os.environ.get("COMSPEC", "cmd.exe")
+                # gh-101283: without a fully-qualified path, before Windows
+                # checks the system directories, it first looks in the
+                # application directory, and also the current directory if
+                # NeedCurrentDirectoryForExePathW(ExeName) is true, so try
+                # to avoid executing unqualified "cmd.exe".
+                comspec = os.environ.get('ComSpec')
+                if not comspec:
+                    system_drive = os.environ.get('SystemDrive') or 'C:'
+                    system_root = os.environ.get('SystemRoot') or os.path.join(
+                        system_drive, 'Windows')
+                    comspec = os.path.join(system_root, 'System32', 'cmd.exe')
+                    if not os.path.isfile(comspec):
+                        # cmd.exe is missing, or the system environment
+                        # variables are broken, or they're undefined and the
+                        # system is installed into a non-standard location.
+                        # This is highly unlikely, and we cannot help here.
+                        comspec = 'cmd.exe'
+                if not executable and os.path.isabs(comspec):
```

Comment:
This is a long-standing platform inconsistency. On POSIX, combining `executable` with `shell=True` replaces the shell path in `args`. It should have been the same on Windows. Instead we allow an arbitrary `exectuable` to be mismatched with a command-line based on `ComSpec`. I wouldn't want that behavior to be documented. I'd prefer to eliminate the inconsistency in 3.12 and document consistent, sensible behavior.

If the caller didn't pass `executable`, I'm suggesting to use it in order to speed up the `CreateProcessW()` call a bit. When `lpApplicationName` (i.e. `executable`) is `NULL`, `CreateProcessW()` has to parse the command name out of the command line and call `SearchPathW(). That step gets skipped if `lpApplicationName` isn't `NULL`, in which case the path name of the executable gets resolved against the current working directory.

Reply:
It's basically already documented today that `executable` will override the launched process but not the command line arguments (I've actually used that deliberately for testing commands that inspect their own argv[0]), so all I'm proposing is clarifying that specifying it with `shell=True` will still do that and it may not be a good idea.

I sure don't want to start messing with the command line based on `executable`. It's so useful

Reply:
> all I'm proposing is clarifying that specifying it with `shell=True` will still do that and it may not be a good idea.

I'd rather fix it in 3.12 and document the correct behavior. On POSIX with `shell=True`, using `executable` replaces the shell in the command line, which is the sensible thing to do. On Windows, that would entail using `executable` instead of the value of the `ComSpec` environment variable, but that was never implemented.

Reply:
Actually yeah, that makes sense. Somehow I blanked out that we add the first argument in ourselves anyway.